impl/src/main/java/org/jboss/weld/contexts/CreationalContextImpl.java

```diff
@@ -139,8 +150,33 @@ public void release(Contextual<T> contextual, T instance) {
         }
     }
 
-    private static <T> void destroy(ContextualInstance<T> beanInstance) {
-        beanInstance.getContextual().destroy(beanInstance.getInstance(), beanInstance.getCreationalContext());
+    private <T> void destroy(ContextualInstance<T> beanInstance) {
+        // Precondition: synchronized (dependentInstances)
+        if (this.destroyed.add(beanInstance)) {
+            beanInstance.getContextual().destroy(beanInstance.getInstance(), beanInstance.getCreationalContext());
+        }
+    }
+
+    private <T> void release(ContextualInstance<T> beanInstance) {
+        // Precondition: synchronized (dependentInstances)
+        if (this.destroyed.add(beanInstance)) {
+            CreationalContext<T> cc = beanInstance.getCreationalContext();
+            if (cc instanceof CreationalContextImpl<?>) {
+                ((CreationalContextImpl<T>) cc).release(beanInstance.getContextual(), beanInstance.getInstance());
+            } else {
+                cc.release();
+            }
+        }
+    }
+
+    private boolean isDestroyed(Contextual<?> c) {
```

Comment:
IDE shows me this method is never really used - probably a leftover from your earlier iterations?

Reply:
Done